src/main/scala/io/iohk/ethereum/network/discovery/PeerDiscoveryManager.scala

```diff
@@ -35,16 +59,21 @@ class PeerDiscoveryManager(
         else
           knownNodesStorage.getKnownNodes().map(Node.fromUri)
 
-      bootstrapNodes ++ knownNodes
+      (bootstrapNodes ++ knownNodes).filterNot(isLocalNode).toVector
     }
 
   override def receive: Receive = init
 
-  // The service hasn't been started yet, so it just serves the static known nodes.
-  def init: Receive = {
+  private def handleNodeInfoRequests(discovery: Option[Discovery]): Receive = {
     case GetDiscoveredNodesInfo =>
-      sendDiscoveredNodesInfo(None, sender)
+      sendDiscoveredNodesInfo(discovery.map(_._1), sender)
+
+    case GetRandomNodeInfo =>
+      sendRandomNodeInfo(discovery.map(_._2), sender)
```

Comment:
I am not well versed in Iterant so I may be missing something, but doesn't such strategy means that we will have muliple concurrent lookups on going ? i.e
1. PeerManager make connection to 10 peers
2. All of those connections fails
3. PeerDiscoveryManager receivess 10 GetRandomNodeInfo messages which means it starts 10 conncurent random lookupus ?

Reply:
The reason this is not going to be happening is because those 10 requests are not doing 10 lookups, they are trying to pull 10 items from the [concurrent consumer](https://github.com/input-output-hk/mantis/pull/803/files?file-filters%5B%5D=.nix&file-filters%5B%5D=.scala#diff-b8b5b45fbe257d7ed9e0ff425a34704bc1939df3faab5accc1b28c8413a1f551R41-R44) which is based on the iterant and has its own internal queue. 

What's [happening](https://github.com/monix/monix/blob/series/3.x/monix-tail/shared/src/main/scala/monix/tail/internal/IterantConsume.scala) is that we set up a consumer with a limited capacity, into which the iterant is constantly trying to push its items, but it pauses when the capacity is full. The [pull](https://github.com/input-output-hk/mantis/pull/803/files?file-filters%5B%5D=.nix&file-filters%5B%5D=.scala#diff-b8b5b45fbe257d7ed9e0ff425a34704bc1939df3faab5accc1b28c8413a1f551R198) which the requests kick off just tries to take 1 items from the underlying queue, which frees up space for the iterant to fill, which may result in another lookup happening via `repeatEvalF`. 

The good thing about the the [ConcurrentChannel](https://monix.io/api/3.0/monix/catnap/ConcurrentChannel.html) used by `Iterant.consumeWithConfig` is that it can be consumed concurrently and it replicates messages to all consumers as well, which is what I ended up using in [Scalanet ReqResponseChannel](https://github.com/input-output-hk/scalanet/pull/106/files#diff-8fe185f0be9efda7b0d43c69976d57d2ac5b0cb330c44f5d240057dda7fd3e67R182-R195) to replace the `ConnectableSubject`.

Reply:
So with a buffer size of 45 and a bucket size of 16 what will happen (or rather, what _would_ happen if the lookups were infinitely fast) is that the iterant will do a lookup, get 16 peers, push them into the consumer, which will have 29 more spaces in the buffer, so the iterant will do 2 more lookups to fill those, and ending up blocked trying to push the last 3 items it has. The first 2 pulls will not result in further lookups, but after pulling the 3rd item the iterant will again do a lookup and be blocked trying to push the next 16 items into the queue.

Reply:
oh sounds neat! Could you maybe add this (or some shortened version) as a commment where Iterant is instantiated ?
Most of the people (including me) are pretty well versed with observable and task, and Iterant is somewhat less used part of monix, and it would be nice to have this expalantion.

Reply:
Sure, added comments. I liked the `Iterant` a lot, it's very similar to the `Observable` in its API, but constructing it recursively is much more intuitive, so things like lazy database iterators using cursors are easier to build.